src/coreclr/jit/codegencommon.cpp

```diff
@@ -12528,3 +12534,83 @@ void CodeGenInterface::VariableLiveKeeper::dumpLvaVariableLiveRanges() const
 }
 #endif // DEBUG
 #endif // USING_VARIABLE_LIVE_RANGE
+
+void CodeGen::genSetRegsModifiedForPoisoning()
+{
+    // For ARM32 we need a callee-reserved register since the scratch registers may have args.
+#if defined(TARGET_ARM)
+    regSet.rsSetRegsModified(RBM_R4);
+#endif
+}
+
+//-----------------------------------------------------------------------------
+// genPoisonFrame: Generate code that places a recognizable value into address exposed variables.
+//
+// Remarks:
+//   This function emits code to poison address exposed non-zero-inited local variables. We expect this function
+//   to be called when emitting code for the scratch BB that comes right after the prolog.
+//   The variables are poisoned using 0xcccccccc. This value is stored in EAX/RAX on xarch and r/x0 on ARM,
+//   so we expect that this reg is not live.
+void CodeGen::genPoisonFrame(regMaskTP regLiveIn)
+{
+#if defined(TARGET_XARCH)
+    const regNumber immRegNum = REG_EAX;
+#elif defined(TARGET_ARM)
+    const regNumber immRegNum = REG_R4;
+#elif defined(TARGET_ARM64)
+    const regNumber immRegNum = REG_R9;
+#else
+    return;
+#endif
+
+    assert(compiler->compShouldPoisonFrame());
+    assert((regLiveIn & genRegMask(immRegNum)) == 0);
+
+    // The first time we need to poison something we will initialize a register to the largest immediate cccccccc that we can fit.
+    bool hasPoisonImm = false;
+    for (unsigned varNum = 0; varNum < compiler->info.compLocalsCount; varNum++)
+    {
+        LclVarDsc* varDsc = compiler->lvaGetDesc(varNum);
+        if (varDsc->lvIsParam || varDsc->lvMustInit || !varDsc->lvAddrExposed)
+        {
+            continue;
+        }
+
+        assert(varDsc->lvOnFrame);
+
+        if (!hasPoisonImm)
+        {
+#ifdef TARGET_64BIT
+            instGen_Set_Reg_To_Imm(EA_8BYTE, immRegNum, (ssize_t)0xcccccccccccccccc);
+#else
+            instGen_Set_Reg_To_Imm(EA_4BYTE, immRegNum, (ssize_t)0xcccccccc);
+#endif
+            hasPoisonImm = true;
+        }
+
+        // For 64-bit we check if the local is 8-byte aligned. For 32-bit, we assume everything is always 4-byte aligned.
+#ifdef TARGET_64BIT
+        bool fpBased;
+        int addr = compiler->lvaFrameAddress((int)varNum, &fpBased);
+#else
+        int addr = 0;
+#endif
+        int size = (int)compiler->lvaLclSize(varNum);
+        int end = addr + size;
+        for (int offs = addr; offs < end;)
+        {
+#ifdef TARGET_64BIT
+            if ((offs & 7) == 0 && end - offs >= 8)
```

Comment:
Nit: I would prefer to avoid  unnamed constants, maybe similar to what we do in `genCodeForCpBlkUnroll` https://github.com/dotnet/runtime/blob/1d84e014ac2b6d11b1ddc452ce1a7706395abcb6/src/coreclr/jit/codegenarmarch.cpp#L2206 
```
const unsigned longRegSize = genTypeSize(TYP_LONG);
if ((offs % longRegSize  == 0) && (end - offs >= longRegSize))
...
offs += longRegSize;
...
```

Reply:
IMO, these constants are so fundamental that there is not much to be gained doing that. `EA_8BYTE` is just as unnamed as `8` in my opinion. But I can change it if you still think it's better.

Reply:
I switched to modulo but kept `4` and `8`.